File: Blind 75/07 - Trees/98-validate-binary-search-tree.py
'''
Given the root of a binary tree, determine if it is a valid binary search tree (BST).

A valid BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.
'''
from math import inf
from time import time
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def valid(self, node, left, right):
        if not node:
            return True
        if not (left < node.val < right):
            return False

        # Pass down parent value as well
        return self.valid(node.left, left, node.val) and self.valid(
            node.right, node.val, right
        )

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.valid(root, -inf, inf)

    def reference(self, root: Optional[TreeNode]) -> bool:
        return self.valid(root, float('-inf'), float('inf'))

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.isValidBST(case))
                else:
                    self.isValidBST(case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        TreeNode(2, TreeNode(1), TreeNode(3)),
        TreeNode(5, TreeNode(1), TreeNode(4, TreeNode(3), TreeNode(6))),
        # Additional
        TreeNode(),
        TreeNode(1, TreeNode(1)),
    ]
    test.quantify(test_cases)

File: Blind 75/07 - Trees/JavaScript/98-validate-binary-search-tree.js
const { performance } = require('perf_hooks');

class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class Solution {
  method(root) {
    const dfs = (node, left, right) => {
      if (!node) return true;
      if (left >= node.val || right <= node.val) return false;

      return dfs(node.left, left, node.val) && dfs(node.right, node.val, right);
    };

    return dfs(root, -Infinity, Infinity);
  }

  reference(root, min = -Infinity, max = Infinity) {
    const isBaseCase = root === null;
    if (isBaseCase) return true;

    const isInvalid = root.val <= min || max <= root.val;
    if (isInvalid) return false;

    return this.dfs(root, min, max);
  }

  dfs = (root, min, max) => {
    const left = this.reference(root.left, min, root.val);
    const right = this.reference(root.right, root.val, max);

    return left && right;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  new TreeNode(2, new TreeNode(1), new TreeNode(3)),
  new TreeNode(
    5,
    new TreeNode(1),
    new TreeNode(4, new TreeNode(3), new TreeNode(6))
  ),
  // Additional
  new TreeNode(),
  new TreeNode(1, new TreeNode(1)),
];
test.quantify(testCases);
